cpp/daal/src/algorithms/k_nearest_neighbors/bf_knn_impl.i

```diff
@@ -0,0 +1,309 @@
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef __BF_KNN_IMPL_I__
+#define __BF_KNN_IMPL_I__
+
+#include "algorithms/engines/engine.h"
+#include "services/daal_defines.h"
+
+#include "algorithms/k_nearest_neighbors/bf_knn_classification_model.h"
+#include "src/algorithms/k_nearest_neighbors/bf_knn_classification_train_kernel.h"
+#include "src/algorithms/k_nearest_neighbors/bf_knn_classification_predict_kernel.h"
+#include "src/algorithms/k_nearest_neighbors/oneapi/bf_knn_classification_model_ucapi_impl.h"
+
+#include "data_management/data/numeric_table.h"
+#include "services/env_detect.h"
+#include "src/services/service_data_utils.h"
+#include "src/services/service_utils.h"
+#include "src/services/service_defines.h"
+#include "src/threading/threading.h"
+#include "src/data_management/service_numeric_table.h"
+
+#include "src/algorithms/service_kernel_math.h"
+#include "src/algorithms/service_sort.h"
+
+#include "src/externals/service_math.h"
+
+namespace daal
+{
+namespace algorithms
+{
+namespace bf_knn_classification
+{
+namespace internal
+{
+const size_t classBufSize = 10;
+
+template <typename FPType, CpuType cpu>
+class BruteForceNearestNeighbors
+{
+public:
+    BruteForceNearestNeighbors(char metric) : _metric(metric) {}
+
+    ~BruteForceNearestNeighbors()
+    {
+        if (_distances)
+        {
+            services::internal::service_free<FPType, cpu>(_distances);
+        }
+    }
+
+    services::Status kNearest(const size_t k, int * neighborsIndices, FPType * neighborsDistances, const NumericTable * trainTable,
+                              const NumericTable * testTable)
+    {
+        computeDistances(trainTable, testTable);
+
+        const size_t nTrain = trainTable->getNumberOfRows();
+        const size_t nTest  = testTable->getNumberOfRows();
+
+        const size_t outerBlockSize = 32;
+        const size_t nOuterBlocks   = nTest / outerBlockSize + !!(nTest % outerBlockSize);
+
+        daal::threader_for(nOuterBlocks, nOuterBlocks, [&](size_t outerBlock) {
+            const size_t outerStart = outerBlock * outerBlockSize;
+            const size_t outerEnd   = outerBlock + 1 == nOuterBlocks ? nTest : outerStart + outerBlockSize;
+
+            for (size_t i = outerStart; i < outerEnd; ++i)
+            {
+                daal::services::internal::TArray<int, cpu> indicesArr(nTrain);
+                int * indices = indicesArr.get();
+                for (size_t j = 0; j < nTrain; ++j)
+                {
+                    indices[j] = j;
+                }
+
+                daal::algorithms::internal::qSort<FPType, int, cpu>(nTrain, _distances + i * nTrain, indices);
```

Comment:
What about using heap to select neighbors? In general it should provide better performance since:
1) The complexity is $O(N \cdot \log_2 (K))$ instead of **average** complexity of $O(N \cdot \log-2 (N))$ for quick sort
2) For small K this heap can be placed in CPU cache

Reply:
Extra optimizations will be added in future updates, but thanks for comment.